pkg/component/crddeployer/crd.go

```diff
@@ -0,0 +1,155 @@
+	var fns []flow.TaskFn
+
+	for resourceName, _ := range c.crdNameToManifest {
+		fns = append(fns, func(ctx context.Context) error {
+			crd := &apiextensionsv1.CustomResourceDefinition{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: resourceName,
+				},
+			}
+			return client.IgnoreNotFound(c.client.Delete(ctx, crd))
+		})
+	}
+
+	return flow.Parallel(fns...)(ctx)
+}
+
+// Wait waits for the CRDs to be deployed.
+func (c *crdDeployer) Wait(ctx context.Context) error {
+	return WaitUntilCRDManifestsReady(ctx, c.client, maps.Keys(c.crdNameToManifest))
+}
+
+// WaitCleanup waits for destruction to finish and CRDs to be fully removed.
+func (c *crdDeployer) WaitCleanup(ctx context.Context) error {
+	return WaitUntilCRDManifestsDestroyed(ctx, c.client, maps.Keys(c.crdNameToManifest))
+}
+
+// MakeCRDNameMap returns a map that has the name of the resource as key, and the corresponding manifest as value.
+func MakeCRDNameMap(manifests []string) (map[string]string, error) {
+	crdNameToManifest := make(map[string]string)
+	for _, manifest := range manifests {
+		name, err := GetObjectNameFromManifest(manifest)
+		if err != nil {
+			return nil, err
+		}
+		crdNameToManifest[name] = manifest
+	}
+	return crdNameToManifest, nil
+}
+
+// WaitUntilCRDManifestsReady takes names of CRDs and waits for them to get ready with a timeout of 15 seconds.
+func WaitUntilCRDManifestsReady(ctx context.Context, c client.Client, crdNames []string) error {
+	var fns []flow.TaskFn
+	for _, crdName := range crdNames {
+		fns = append(fns, func(ctx context.Context) error {
+			timeoutCtx, cancel := context.WithTimeout(ctx, CRDWaitTimeout)
+			defer cancel()
+
+			return retry.Until(timeoutCtx, 1*time.Second, func(ctx context.Context) (done bool, err error) {
+				crd := &apiextensionsv1.CustomResourceDefinition{}
+
+				if err := c.Get(ctx, client.ObjectKey{Name: crdName}, crd); err != nil {
+					return retry.SevereError(err)
+				}
+
+				if err := health.CheckCustomResourceDefinition(crd); err != nil {
+					return retry.MinorError(err)
+				}
+				return retry.Ok()
+			})
+		})
+	}
+	return flow.Parallel(fns...)(ctx)
+}
+
+// WaitUntilCRDManifestsDestroyed takes names of CRDs and waits for them to get destroted with a timeout of 15 seconds.
+func WaitUntilCRDManifestsDestroyed(ctx context.Context, c client.Client, crdNames []string) error {
+	var fns []flow.TaskFn
+
+	for _, resourceName := range crdNames {
+		crd := &apiextensionsv1.CustomResourceDefinition{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: resourceName,
+			},
+		}
+
+		fns = append(fns, func(ctx context.Context) error {
+			timeoutCtx, cancel := context.WithTimeout(ctx, CRDWaitTimeout)
+			defer cancel()
+			return kubernetesutils.WaitUntilResourceDeleted(timeoutCtx, c, crd, 1*time.Second)
+		})
+	}
+	return flow.Parallel(fns...)(ctx)
+}
+
+// GetObjectNameFromManifest takes a manifest and returns its corresponding name.
+func GetObjectNameFromManifest(manifest string) (string, error) {
```

Comment:
This function is not really needed when I saw correctly in https://github.com/gardener/gardener/pull/10639. Could you please drop it? This `crd` package would anyway not be the perfect place for such a generic function.

Reply:
I used it [here](https://github.com/gardener/gardener/pull/10639/files#diff-a21beedc94f33d1eb00d113858fa12050121bc8e0852520726150b4eda29626aR37) for the special case of the `vpa` CRD deployer.

Will drop it and possibly reintroduce (at a suitable spot then) when needed in the future PR

Reply:
Thanks. 

Side remark: It's used at the linked place, but var `resourceNames` doesn't really have a meaning in the `vpa` package.